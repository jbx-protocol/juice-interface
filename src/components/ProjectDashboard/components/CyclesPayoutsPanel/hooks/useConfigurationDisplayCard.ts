import { t } from '@lingui/macro'
import { useMemo } from 'react'

export const useConfigurationDisplayCard = (type: 'current' | 'upcoming') => {
  const title = useMemo(() => {
    if (type === 'current') {
      return t`Current cycle`
    }
    return t`Upcoming cycle`
  }, [type])

  return {
    title,
  }
}
