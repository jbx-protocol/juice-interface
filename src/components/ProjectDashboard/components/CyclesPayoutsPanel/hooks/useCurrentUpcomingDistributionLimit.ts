import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
