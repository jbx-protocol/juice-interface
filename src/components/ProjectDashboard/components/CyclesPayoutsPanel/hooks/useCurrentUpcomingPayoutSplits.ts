import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { ETH_PAYOUT_SPLIT_GROUP } from 'constants/splits'
import useProjectSplits from 'hooks/v2v3/contractReader/useProjectSplits'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingPayoutSplits = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const { payoutSplits: currentPayoutSplits } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingPayoutSplits, loading: upcomingProjectSplitsLoading } =
    useProjectSplits({
      projectId,
      splitGroup: ETH_PAYOUT_SPLIT_GROUP,
      domain: upcomingFundingCycle?.configuration?.toString(),
    })

  if (type === 'current') {
    return { splits: currentPayoutSplits, loading: false }
  }
  return {
    splits: upcomingPayoutSplits,
    loading: upcomingFundingCycleLoading || upcomingProjectSplitsLoading,
  }
}
