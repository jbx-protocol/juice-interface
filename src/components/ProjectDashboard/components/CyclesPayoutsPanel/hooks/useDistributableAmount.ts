import { V2V3ProjectContext } from 'contexts/v2v3/Project/V2V3ProjectContext'
import { BigNumber } from 'ethers'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { useContext } from 'react'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'

export const useDistributableAmount = () => {
  const {
    distributionLimit,
    usedDistributionLimit,
    balanceInDistributionLimitCurrency,
    distributionLimitCurrency,
  } = useContext(V2V3ProjectContext)

  const currency =
    distributionLimitCurrency?.toNumber() ??
    (V2V3_CURRENCY_ETH as V2V3CurrencyOption)

  const effectiveDistributionLimit = distributionLimit ?? BigNumber.from(0)
  const distributedAmount = usedDistributionLimit ?? BigNumber.from(0)
  const balInDistributionLimit =
    balanceInDistributionLimitCurrency ?? BigNumber.from(0)

  const distributable = effectiveDistributionLimit.eq(0)
    ? effectiveDistributionLimit
    : effectiveDistributionLimit.sub(distributedAmount)

  const distributableAmount = balInDistributionLimit.gt(distributable)
    ? distributable
    : balInDistributionLimit

  return {
    distributableAmount,
    currency,
  }
}
