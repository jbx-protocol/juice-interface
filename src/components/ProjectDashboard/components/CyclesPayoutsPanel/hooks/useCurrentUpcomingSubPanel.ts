import { t } from '@lingui/macro'
import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { useFundingCycleCountdown } from 'components/ProjectDashboard/hooks/useFundingCycleCountdown'
import { timeSecondsToDateString } from 'components/ProjectDashboard/utils/timeSecondsToDateString'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { useMemo } from 'react'

export const useCurrentUpcomingSubPanel = (type: 'current' | 'upcoming') => {
  const { projectId } = useProjectMetadata()
  const {
    fundingCycle,
    loading: { fundingCycleLoading },
  } = useProjectContext()
  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []
  const { timeRemainingText } = useFundingCycleCountdown()

  const cycleNumber = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.number.toNumber()
    }
    return fundingCycle?.number ? fundingCycle.number.toNumber() + 1 : undefined
  }, [fundingCycle?.number, type])

  const cycleUnlocked = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.duration?.isZero() ?? true
    }
    return upcomingFundingCycle?.duration?.isZero() ?? true
  }, [fundingCycle?.duration, type, upcomingFundingCycle?.duration])

  const upcomingCycleLength = useMemo(() => {
    if (!upcomingFundingCycle) return
    if (cycleUnlocked) return '-'
    return timeSecondsToDateString(
      upcomingFundingCycle.duration.toNumber(),
      'short',
    )
  }, [cycleUnlocked, upcomingFundingCycle])

  const status = useMemo(() => {
    if (cycleUnlocked) return t`Unlocked`
    return t`Locked`
  }, [cycleUnlocked])

  const remainingTime = useMemo(() => {
    if (cycleUnlocked) return '-'
    return timeRemainingText
  }, [cycleUnlocked, timeRemainingText])

  // Short circuit current for faster loading
  if (type === 'current') {
    if (fundingCycleLoading) return { loading: true, type }
    return {
      loading: false,
      type,
      cycleNumber,
      status,
      remainingTime,
    }
  }

  if (fundingCycleLoading || upcomingFundingCycleLoading)
    return {
      loading: true,
      type,
    }

  return {
    loading: false,
    type,
    cycleNumber,
    status,
    cycleLength: upcomingCycleLength,
  }
}
