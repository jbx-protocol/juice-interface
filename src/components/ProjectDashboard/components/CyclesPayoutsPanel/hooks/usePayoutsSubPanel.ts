import { useProjectContext } from 'components/ProjectDashboard/hooks'
import { BigNumber } from 'ethers'
import { Split } from 'models/splits'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { useCallback, useMemo } from 'react'
import assert from 'utils/assert'
import { fromWad } from 'utils/format/formatNumber'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'
import { getProjectOwnerRemainderSplit } from 'utils/splits'
import { isJuiceboxProjectSplit } from 'utils/v2v3/distributions'
import {
  MAX_DISTRIBUTION_LIMIT,
  SPLITS_TOTAL_PERCENT,
  feeForAmount,
  formatSplitPercent,
} from 'utils/v2v3/math'
import { useCurrentUpcomingDistributionLimit } from './useCurrentUpcomingDistributionLimit'
import { useCurrentUpcomingPayoutSplits } from './useCurrentUpcomingPayoutSplits'

const splitHasFee = (split: Split) => {
  return !isJuiceboxProjectSplit(split)
}

const calculateSplitAmountWad = (
  split: Split,
  distributionLimit: BigNumber | undefined,
  primaryETHTerminalFee: BigNumber | undefined,
) => {
  const splitValue = distributionLimit
    ?.mul(split.percent)
    .div(SPLITS_TOTAL_PERCENT)
  const feeAmount = splitHasFee(split)
    ? feeForAmount(splitValue, primaryETHTerminalFee) ?? BigNumber.from(0)
    : BigNumber.from(0)
  return splitValue?.sub(feeAmount)
}

export const usePayoutsSubPanel = (type: 'current' | 'upcoming') => {
  const { splits, loading } = useCurrentUpcomingPayoutSplits(type)
  const { projectOwnerAddress, primaryETHTerminalFee } = useProjectContext()

  const { distributionLimit, distributionLimitCurrency } =
    useCurrentUpcomingDistributionLimit(type)

  const showAmountOnPayout = useMemo(() => {
    if (distributionLimit?.eq(MAX_DISTRIBUTION_LIMIT)) return false
    if (distributionLimit?.eq(0)) return false
    return true
  }, [distributionLimit])

  const transformSplit = useCallback(
    (split: Split) => {
      assert(split.beneficiary, 'Beneficiary must be defined')
      let amount = undefined
      const splitAmountWad = calculateSplitAmountWad(
        split,
        distributionLimit,
        primaryETHTerminalFee,
      )
      if (showAmountOnPayout && splitAmountWad && distributionLimitCurrency) {
        amount = formatCurrencyAmount({
          amount: Number(fromWad(splitAmountWad)),
          currency: distributionLimitCurrency.toNumber() as V2V3CurrencyOption,
        })
      }
      return {
        projectId: split.projectId
          ? BigNumber.from(split.projectId).toNumber()
          : undefined,
        address: split.beneficiary!,
        percent: `${formatSplitPercent(BigNumber.from(split.percent))}%`,
        amount,
      }
    },
    [
      distributionLimit,
      distributionLimitCurrency,
      primaryETHTerminalFee,
      showAmountOnPayout,
    ],
  )

  const totalPayoutAmount = useMemo(() => {
    if (!distributionLimit || !distributionLimitCurrency) return
    if (
      distributionLimit.eq(MAX_DISTRIBUTION_LIMIT) ||
      distributionLimit.isZero()
    )
      return
    return formatCurrencyAmount({
      amount: fromWad(distributionLimit),
      currency: distributionLimitCurrency.toNumber() as V2V3CurrencyOption,
    })
  }, [distributionLimit, distributionLimitCurrency])

  const payouts = useMemo(() => {
    if (loading || !splits) return

    const ownerPayout = projectOwnerAddress
      ? getProjectOwnerRemainderSplit(projectOwnerAddress, splits)
      : undefined

    return [
      ...splits,
      ...(ownerPayout && ownerPayout.percent > 0 ? [ownerPayout] : []),
    ]
      .sort((a, b) => Number(b.percent) - Number(a.percent))
      .map(transformSplit)
  }, [loading, projectOwnerAddress, splits, transformSplit])

  return {
    loading,
    payouts,
    totalPayoutAmount,
  }
}
