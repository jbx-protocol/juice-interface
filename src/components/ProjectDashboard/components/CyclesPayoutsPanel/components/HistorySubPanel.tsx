import { Disclosure, Transition } from '@headlessui/react'
import { ChevronDownIcon } from '@heroicons/react/24/outline'
import { Trans, t } from '@lingui/macro'
import {
  V2V3FundingCycle,
  V2V3FundingCycleMetadata,
} from 'models/v2v3/fundingCycle'
import { Fragment } from 'react'
import { twMerge } from 'tailwind-merge'
import { useHistorySubPanel } from '../hooks/useHistorySubPanel'
import { HistoricalConfigurationPanel } from './HistoricalConfigurationPanel'

export type HistoryData = {
  _metadata: {
    fundingCycle: V2V3FundingCycle
    metadata: V2V3FundingCycleMetadata
  }
  cycleNumber: string
  withdrawn: string
  date: string
}[]

export const HistorySubPanel = () => {
  const { loading, data, error } = useHistorySubPanel()
  const tableHeaders = [t`Cycle #`, t`Withdrawn`, t`Date`]

  return data.length || loading ? (
    <div className="grid min-w-full grid-cols-1">
      <div className="rounded-t-lg bg-smoke-50 p-4 pr-2 dark:bg-slate-700">
        <div className="grid grid-cols-config-table gap-3">
          {tableHeaders.map(header => (
            <div
              key={header}
              className={twMerge('text-start text-sm font-semibold')}
            >
              {header}
            </div>
          ))}
          <div className="relative w-6 min-w-[24px] py-2.5">
            <span className="sr-only">View</span>
          </div>
        </div>
      </div>

      <div className="divide-y divide-grey-200 dark:divide-slate-500">
        {loading ? (
          <>
            <SkeletonRow />
            <SkeletonRow />
            <SkeletonRow />
            <SkeletonRow />
          </>
        ) : error ? (
          <div className="text-error-400">{error}</div>
        ) : (
          <>
            {data.map(cycle => (
              <Disclosure key={cycle.cycleNumber} as={Fragment}>
                {({ open }) => (
                  <Disclosure.Button
                    data-testid={`disclosure-button-${cycle.cycleNumber}`}
                    as="div"
                    className="cursor-pointer p-4 pr-2"
                  >
                    <div className="grid grid-cols-config-table gap-3 whitespace-nowrap text-sm font-medium">
                      <div>#{cycle.cycleNumber}</div>
                      <div>{cycle.withdrawn}</div>
                      <div className="text-grey-500 dark:text-slate-200">
                        {cycle.date}
                      </div>
                      <div className="text-gray-500 flex items-center justify-end whitespace-nowrap text-sm">
                        <ChevronDownIcon
                          className={twMerge(open && 'rotate-180', 'h-5 w-5')}
                        />
                      </div>
                    </div>
                    <Transition
                      show={open}
                      as={Fragment}
                      enter="transition-all ease-in-out duration-300"
                      enterFrom="max-h-0 overflow-hidden opacity-0"
                      enterTo="max-h-[1000px] overflow-hidden opacity-100"
                      leave="transition-all ease-in-out duration-300"
                      leaveFrom="max-h-[1000px] overflow-hidden opacity-100"
                      leaveTo="max-h-0 overflow-hidden opacity-0"
                    >
                      <Disclosure.Panel className="p-4">
                        <HistoricalConfigurationPanel {...cycle._metadata} />
                      </Disclosure.Panel>
                    </Transition>
                  </Disclosure.Button>
                )}
              </Disclosure>
            ))}
          </>
        )}
      </div>
    </div>
  ) : (
    <div className="text-grey-500 dark:text-slate-200">
      <Trans>No previous cycles.</Trans>
    </div>
  )
}

const SkeletonRow = () => (
  <div className="p-4 pr-2">
    <div className="grid animate-pulse grid-cols-config-table bg-white dark:bg-slate-950">
      <div className="whitespace-nowrap text-sm font-medium">
        <div className="h-4 w-14 rounded bg-grey-200 dark:bg-slate-500"></div>
      </div>
      <div className="whitespace-nowrap text-sm font-medium">
        <div className="h-4 w-24 rounded bg-grey-200 dark:bg-slate-500"></div>
      </div>
      <div className="whitespace-nowrap text-sm text-grey-500">
        <div className="h-4 w-20 rounded bg-grey-200 dark:bg-slate-500"></div>
      </div>
      <div className="relative w-6 min-w-[24px] py-2.5"></div>
    </div>
  </div>
)
