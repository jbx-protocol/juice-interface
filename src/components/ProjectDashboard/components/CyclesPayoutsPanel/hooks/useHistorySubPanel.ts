import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { fetchPastFundingCycles } from 'components/v2v3/V2V3Project/V2V3FundingCycleSection/FundingCycleHistory/utils'
import { ETH_TOKEN_ADDRESS } from 'constants/v2v3/juiceboxTokens'
import { V2V3ProjectContractsContext } from 'contexts/v2v3/ProjectContracts/V2V3ProjectContractsContext'
import { BigNumber, Contract } from 'ethers'
import { JB_CONTROLLER_V_3_1 } from 'hooks/v2v3/V2V3ProjectContracts/projectContractLoaders/useProjectController'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { V2V3FundingCycle } from 'models/v2v3/fundingCycle'
import moment from 'moment'
import { useCallback, useContext, useEffect, useReducer } from 'react'
import { isBigNumberish } from 'utils/bigNumbers'
import { callContract } from 'utils/callContract'
import { fromWad } from 'utils/format/formatNumber'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'
import { HistoryData } from '../components/HistorySubPanel'

type State = {
  loading: boolean
  data: HistoryData
  error?: string
}

type Action =
  | { type: 'request' }
  | { type: 'success'; payload: HistoryData }
  | { type: 'failure'; payload: string }

const reducer = (state: State, action: Action) => {
  switch (action.type) {
    case 'request':
      return { ...state, loading: true, error: undefined }
    case 'success':
      return { ...state, loading: false, data: action.payload }
    case 'failure':
      return { ...state, loading: false, error: action.payload }
  }
}

export const useHistorySubPanel = () => {
  const { projectId } = useProjectMetadata()
  const { fundingCycle, primaryETHTerminal } = useProjectContext()
  const {
    contracts,
    versions: { JBControllerVersion },
  } = useContext(V2V3ProjectContractsContext)
  const [state, dispatch] = useReducer(reducer, {
    loading: false,
    data: [],
  })

  const fetchCurrencyOfFundingCycle = useCallback(
    async (fundingCycle: V2V3FundingCycle) => {
      const terminal = primaryETHTerminal
      const { JBController, JBFundAccessConstraintsStore } = contracts
      const contract =
        JBControllerVersion === JB_CONTROLLER_V_3_1
          ? JBFundAccessConstraintsStore
          : JBController
      const result = await callContract({
        contract,
        contracts: contracts as Record<string, Contract>,
        functionName: 'distributionLimitOf',
        args:
          projectId && fundingCycle.configuration && terminal && contract
            ? [
                projectId,
                fundingCycle.configuration.toString(),
                terminal,
                ETH_TOKEN_ADDRESS,
              ]
            : null,
      })
      if (typeof result === undefined) return
      if (!(Array.isArray(result) && result.length === 2)) {
        console.error('Unexpected result from distributionLimitOf', result)
        throw new Error('Unexpected result from distributionLimitOf')
      }
      const [, currency] = result
      if (!isBigNumberish(currency)) {
        console.error('Unexpected result from distributionLimitOf', result)
        throw new Error('Unexpected result from distributionLimitOf')
      }
      const currencyOption = BigNumber.from(currency).toNumber()
      if (currencyOption === 0) {
        return undefined
      }
      return BigNumber.from(currency).toNumber() as V2V3CurrencyOption
    },
    [JBControllerVersion, contracts, primaryETHTerminal, projectId],
  )

  const fetchUsedDistributionLimitOfFundingCycle = useCallback(
    async (fundingCycle: V2V3FundingCycle) => {
      const terminal = primaryETHTerminal
      const { number } = fundingCycle
      const result = await callContract({
        contract: contracts.JBETHPaymentTerminalStore,
        contracts: contracts as Record<string, Contract>,
        functionName: 'usedDistributionLimitOf',
        args:
          terminal && projectId && number
            ? [terminal, projectId, number]
            : null,
      })
      if (typeof result === undefined) return
      if (!isBigNumberish(result)) {
        console.error('Unexpected result from usedDistributionLimitOf')
        throw new Error('Unexpected result from usedDistributionLimitOf')
      }
      return Number(fromWad(BigNumber.from(result)))
    },
    [contracts, primaryETHTerminal, projectId],
  )

  const fetchWithdrawnOfFundingCycle = useCallback(
    async (fundingCycle: V2V3FundingCycle) => {
      const [withdrawn, currency] = await Promise.all([
        fetchUsedDistributionLimitOfFundingCycle(fundingCycle),
        fetchCurrencyOfFundingCycle(fundingCycle),
      ])
      return formatCurrencyAmount({ amount: withdrawn, currency }) ?? '0Îž'
    },
    [fetchCurrencyOfFundingCycle, fetchUsedDistributionLimitOfFundingCycle],
  )

  useEffect(() => {
    dispatch({ type: 'request' })

    const fetchHistory = async () => {
      if (!projectId || !fundingCycle || !contracts.JBController) return
      try {
        const data = await fetchPastFundingCycles({
          projectId,
          currentFundingCycle: fundingCycle,
          JBController: contracts.JBController,
        })

        const result = await Promise.all(
          data.map(async d => ({
            cycleNumber: d[0].number.toString(),
            withdrawn: await fetchWithdrawnOfFundingCycle(d[0]),
            date: `${moment(
              d[0].start.add(d[0].duration).mul(1000).toNumber(),
            ).fromNow(true)} ago`,
            _metadata: {
              fundingCycle: d[0],
              metadata: d[1],
            },
          })),
        )

        dispatch({ type: 'success', payload: result })
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } catch (error: any) {
        console.error('error', error)
        dispatch({ type: 'failure', payload: error.message })
      }
    }

    fetchHistory()
  }, [
    contracts.JBController,
    fetchCurrencyOfFundingCycle,
    fetchWithdrawnOfFundingCycle,
    fundingCycle,
    projectId,
  ])

  return state
}
