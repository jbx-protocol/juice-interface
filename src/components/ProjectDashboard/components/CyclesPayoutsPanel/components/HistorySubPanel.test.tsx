/**
 * @jest-environment jsdom
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { useHistorySubPanel } from '../hooks/useHistorySubPanel'
import { HistoryData, HistorySubPanel } from './HistorySubPanel'

jest.mock('../hooks/useHistorySubPanel')
jest.mock('@headlessui/react', () => {
  return {
    __esModule: true,
    Disclosure: jest.requireActual('@headlessui/react').Disclosure,
    Transition: jest
      .fn()
      .mockImplementation(({ children, show }) => (
        <div>{show && children}</div>
      )),
  }
})
jest.mock('./HistoricalConfigurationPanel', () => {
  return {
    __esModule: true,
    HistoricalConfigurationPanel: jest
      .fn()
      .mockImplementation(({ fundingCycle, metadata }) => (
        <div data-testid={fundingCycle.id}>{JSON.stringify(metadata)}</div>
      )),
  }
})

describe('HistorySubPanel', () => {
  const mockData: HistoryData = [
    {
      _metadata: {
        fundingCycle: { id: '1' } as any,
        metadata: { data: 'test data' } as any,
      },
      cycleNumber: '1',
      withdrawn: '100',
      date: '2023-01-01',
    },
  ]

  beforeEach(() => {
    ;(useHistorySubPanel as jest.Mock).mockReturnValue({
      loading: false,
      data: mockData,
      error: null,
    })
  })

  it('renders without crashing', () => {
    render(<HistorySubPanel />)
  })

  it('displays correct cycle data', () => {
    render(<HistorySubPanel />)
    expect(screen.getByText(`#${mockData[0].cycleNumber}`)).toBeInTheDocument()
    expect(screen.getByText(mockData[0].withdrawn)).toBeInTheDocument()
    expect(screen.getByText(mockData[0].date)).toBeInTheDocument()
  })

  it('displays HistoricalConfigurationPanel with correct props on click', async () => {
    render(<HistorySubPanel />)
    const button = screen.getByTestId(
      `disclosure-button-${mockData[0].cycleNumber}`,
    )
    fireEvent.click(button)
    await waitFor(() =>
      expect(
        screen.getByTestId((mockData[0]._metadata.fundingCycle as any).id),
      ).toHaveTextContent(JSON.stringify(mockData[0]._metadata.metadata)),
    )
  })
})
