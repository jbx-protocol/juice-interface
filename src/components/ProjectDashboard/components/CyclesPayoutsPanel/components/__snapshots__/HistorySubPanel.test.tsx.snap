// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistorySubPanel renders 1`] = `
<div>
  <table
    class="min-w-full divide-y divide-grey-200"
  >
    <thead>
      <tr
        class="bg-smoke-50"
      >
        <th
          class="py-3 pl-4 pr-3 text-start text-sm font-semibold"
          scope="col"
        >
          Cycle #
        </th>
        <th
          class="py-3 pl-4 pr-3 text-start text-sm font-semibold"
          scope="col"
        >
          Withdrawn
        </th>
        <th
          class="py-3 pl-4 pr-3 text-start text-sm font-semibold"
          scope="col"
        >
          Date
        </th>
        <th
          class="relative py-3 pl-4 pr-3"
          scope="col"
        >
          <span
            class="sr-only"
          >
            Edit
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      class="divide-y divide-grey-200"
    >
      <tr
        class="animate-pulse bg-white"
      >
        <td
          class="whitespace-nowrap py-4 pl-4 pr-3 text-sm font-medium"
        >
          <div
            class="h-4 w-3/4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm font-medium"
        >
          <div
            class="h-4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm text-grey-500"
        >
          <div
            class="h-4 w-5/6 rounded bg-grey-200"
          />
        </td>
      </tr>
      <tr
        class="animate-pulse bg-white"
      >
        <td
          class="whitespace-nowrap py-4 pl-4 pr-3 text-sm font-medium"
        >
          <div
            class="h-4 w-3/4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm font-medium"
        >
          <div
            class="h-4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm text-grey-500"
        >
          <div
            class="h-4 w-5/6 rounded bg-grey-200"
          />
        </td>
      </tr>
      <tr
        class="animate-pulse bg-white"
      >
        <td
          class="whitespace-nowrap py-4 pl-4 pr-3 text-sm font-medium"
        >
          <div
            class="h-4 w-3/4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm font-medium"
        >
          <div
            class="h-4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm text-grey-500"
        >
          <div
            class="h-4 w-5/6 rounded bg-grey-200"
          />
        </td>
      </tr>
      <tr
        class="animate-pulse bg-white"
      >
        <td
          class="whitespace-nowrap py-4 pl-4 pr-3 text-sm font-medium"
        >
          <div
            class="h-4 w-3/4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm font-medium"
        >
          <div
            class="h-4 rounded bg-grey-200"
          />
        </td>
        <td
          class="whitespace-nowrap px-4 py-4 text-sm text-grey-500"
        >
          <div
            class="h-4 w-5/6 rounded bg-grey-200"
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
